examples/cloudml-collaborative-filtering/preprocessing/preprocess.py

```diff
@@ -0,0 +1,205 @@
+      constants.TFT_TAGS_KEY: tft.compute_and_apply_vocabulary(
+          raw_data[constants.TAGS_KEY],
+          vocab_filename=constants.TAG_VOCAB_NAME,
+          default_value=constants.TFT_DEFAULT_ID),
+      constants.TFT_TOP_10_KEY: tft.apply_vocabulary(
+          raw_data[constants.TOP_10_KEY],
+          item_vocab,
+          default_value=constants.TFT_DEFAULT_ID),
+  }
+  features.update(tft_features)
+  return features
+
+
+def _run_tft_fn(raw_data, tft_fn, transform_fn_path, user_freq, item_freq):
+  """Applys the TensorFlow Transform function to the given data.
+
+  Args:
+    raw_data: a dict of shape {$user_key: $user_id, $item_key: ...}.
+    tft_fn: a TensorFlow Transform function.
+    transform_fn_path: the location to save transformation outputs to.
+    user_freq: minimum frequency of a user to include it in the user vocab.
+    item_freq: minimum frequency of an item to include it in the item vocab.
+
+  Returns:
+    A pCollection of dicts, where each dict is an element of raw_data with the
+      preprocess_fn applied to it:
+      {$user_key: $user_id, $item_key: $item_id, $count_key: $count}.
+  """
+  raw_data_metadata = tft.tf_metadata.dataset_metadata.DatasetMetadata(
+      tft.tf_metadata.dataset_schema.from_feature_spec(constants.TRAIN_SPEC))
+  transformed_dataset, transform_fn = (
+      (raw_data, raw_data_metadata)
+      | beam_impl.AnalyzeAndTransformDataset(
+          lambda x: tft_fn(x, user_freq, item_freq)))
+  _ = (transform_fn
+       | "WriteTransformFn" >>
+       tft.beam.tft_beam_io.transform_fn_io.WriteTransformFn(
+           os.path.join(transform_fn_path, "transform_fn")))
+  return transformed_dataset[0]
+
+
+def _filter_data(features):
+  """Yields samples with valid users or items and then drops tft-created ids."""
+  user_id = features[constants.TFT_USER_KEY]
+  item_id = features[constants.TFT_ITEM_KEY]
+  if (user_id != constants.TFT_DEFAULT_ID
+      and item_id != constants.TFT_DEFAULT_ID):
+    yield {feature: features[feature] for feature in constants.BQ_FEATURES}
+
+
+def _clean_tags(features):
+  if (len(features[constants.TAGS_KEY]) and
+      not features[constants.TAGS_KEY][0]):
+    features[constants.TAGS_KEY] = []
+  return features
+
+
+def _split_data(examples, train_fraction=constants.TRAIN_SIZE,
+                eval_fraction=constants.EVAL_SIZE):
+  """Shuffles and splits the data into train/validation/test."""
+
+  def partition_fn(data, n_partition):  # pylint: disable=unused-argument
+    random_value = np.random.random()
+    if random_value < train_fraction:
+      return 0
+    if random_value < train_fraction + eval_fraction:
+      return 1
+    return 2
+
+  examples_split = examples | "SplitData" >> beam.Partition(partition_fn, 3)
+  return zip(["Train", "Eval", "Test"], examples_split)
+
+
+@beam.ptransform_fn
+def Shuffle(p):
+  """Shuffles the given pCollection."""
+  return (p
+          | "PairWithRandom" >> beam.Map(lambda x: (np.random.random(), x))
```

Comment:
Even though the partitions are made randomly, I believe order within each partition is still preserved. This shuffle makes sure the model is trained with random batches of data. This is important because we want each batch to be representative of the entire dataset, which wouldn't be the case if, say, our input data were seasonal so each batch's data came only from a specific time frame.